### PR TITLE
[Feature][Java/Spring] Support Discriminator Based OneOf Interface

### PR DESCRIPTION
**Pull request by cachescrubber**
# Support Discriminator Based OneOf Interface

**Superseeds**

- https://github.com/OpenAPITools/openapi-generator/pull/10392, which is in turn based on
- https://github.com/OpenAPITools/openapi-generator/pull/8091

**Also Fixes**

- https://github.com/OpenAPITools/openapi-generator/issues/11636

Brings basic discriminator based OneOf Support to the JavaSpring generator. It generates a marker Interface which the oneOf Models then implement. Technically it requires Jackson based models and is implemented using @JsonTypeInfo and @JsonSubTypes based using a discriminator property.

The relevant Code ist moved to AbstractJavaCodegen and aligned with the other Java based generators in the hierarchy.

The oneOf Support is implemented by utilizing the **x-implements**  vendor extension.

Currently It is enabled in the SpringCodegen only:

```
useOneOfInterfaces = true;
legacyDiscriminatorBehavior = false;
``` 

## Synopsis

```
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.EXISTING_PROPERTY, property = "@type", visible = true)
@JsonSubTypes({
  @JsonSubTypes.Type(value = Foo.class, name = "Foo"),
  @JsonSubTypes.Type(value = FooRef.class, name = "FooRef")
})

@Generated(value = "org.openapitools.codegen.languages.SpringCodegen")
public interface FooRefOrValue {
    public String getAtType();
}

```

### PR checklist
 
- [ x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [ x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x ] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [ x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (5.3.0), `6.0.x`
- [ x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

### CODE REVIEW

**Comment by cachescrubber** [comment]
FYI @alexsuperdev @jburgess @wing328

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java [hunk 1]
```diff
@@ -1989,4 +2000,16 @@ protected void addAdditionPropertiesToCodeGenModel(CodegenModel codegenModel, Sc
         }
     }
 
+    @Override
+    public void addImportsToOneOfInterface(List<Map<String, String>> imports) {
+        if (additionalProperties.containsKey(JACKSON)) {
```
Comment (welshm) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java:2005]: Does this mean that GSON does't work?

File: modules/openapi-generator/src/main/resources/JavaSpring/model.mustache [hunk 2]
```diff
@@ -46,7 +46,7 @@ import javax.annotation.Generated;
 {{>enumOuterClass}}
 {{/isEnum}}
 {{^isEnum}}
-{{>pojo}}
+{{#vendorExtensions.x-is-one-of-interface}}{{>oneof_interface}}{{/vendorExtensions.x-is-one-of-interface}}{{^vendorExtensions.x-is-one-of-interface}}{{>pojo}}{{/vendorExtensions.x-is-one-of-interface}}
```
Comment (welshm) [inline modules/openapi-generator/src/main/resources/JavaSpring/model.mustache:49]: No samples use this option currently, so it's hard to evaluate the output currently

File: modules/openapi-generator/src/test/resources/3_0/issue_5381.yaml [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        "@baseType":
+          type: string
+          description: When sub-classing, this defines the super-class
+        "@type":
+          type: string
+          description: When sub-classing, this defines the sub-class Extensible name
+      required:
+        - '@type'
+
+    FooRefOrValue:
```
Comment (welshm) [inline modules/openapi-generator/src/test/resources/3_0/issue_5381.yaml:81]: Given you have a mustache which attempts to determine if it's an interface or class depending on how many nested levels, thoughts on having another container? There is nesting for `allOf` here but not `oneOf`

```
OuterFooRefOrValue:
  type: object
  oneOf:
    - $ref: "#/components/schemas/FooRefOrValue"
    - $ref: "#/components/schemas/AnotherRefOrValue"
  discriminator:
    propertyName: "@type"
      mapping:
        FooRefOrValue: "#/components/schemas/FooRefOrValue"
        AnotherRefOrValue: "#/components/schemas/AnotherRefOrValue"
```

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java (hunk 1 above)
Comment (cachescrubber) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java:2005]: Yes. There is a GSON implementation in okhttp-gson-nextgen and another java based generator. Does the spring generator supports GSON?

File: modules/openapi-generator/src/main/resources/JavaSpring/model.mustache (hunk 2 above)
Comment (cachescrubber) [reply modules/openapi-generator/src/main/resources/JavaSpring/model.mustache:49]: Good point - will add a sample project

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java (hunk 1 above)
Comment (welshm) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java:2005]: I think there is a configuration option for it, but I'm not sure how widely used ti it

File: samples/openapi3/server/petstore/spring-boot-oneof/src/main/java/org/openapitools/model/FooRef.java [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+import java.util.*;
+import javax.annotation.Generated;
+
+/**
+ * FooRef
+ */
+
+
+@Generated(value = "org.openapitools.codegen.languages.SpringCodegen")
+public class FooRef extends EntityRef implements FooRefOrValue {
```
Comment (cachescrubber) [inline samples/openapi3/server/petstore/spring-boot-oneof/src/main/java/org/openapitools/model/FooRef.java:27]: Example: Inherit from EntityRef class using discriminator an implementing the oneOf Interface.

File: samples/openapi3/server/petstore/spring-boot-oneof/src/main/java/org/openapitools/model/FooRefOrValue.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+
+@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.EXISTING_PROPERTY, property = "@type", visible = true)
+@JsonSubTypes({
+  @JsonSubTypes.Type(value = Foo.class, name = "Foo"),
+  @JsonSubTypes.Type(value = FooRef.class, name = "FooRef")
+})
+
+@Generated(value = "org.openapitools.codegen.languages.SpringCodegen")
+public interface FooRefOrValue {
```
Comment (cachescrubber) [inline samples/openapi3/server/petstore/spring-boot-oneof/src/main/java/org/openapitools/model/FooRefOrValue.java:30]: The OneOf Interface

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/JavaClientCodegen.java [hunk 6]
```diff
@@ -896,9 +896,6 @@ public CodegenModel fromModel(String name, Schema model) {
             if (this.parcelableModel) {
                 ((ArrayList<String>) cm.getVendorExtensions().get("x-implements")).add("Parcelable");
             }
-            if (this.serializableModel) {
```
Comment (welshm) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/JavaClientCodegen.java:899]: Was this intentionally removed?

**Review by welshm — COMMENTED** [review]
Makes sense overall - I do wonder how nested `oneOf` will work

**Comment by jburgess** [comment]
@cachescrubber Thank you so much for picking this up and keeping it up-to-date with the base branch! Initial testing looks good on this PR.

@wing328 Can we get this in a milestone?

File: modules/openapi-generator/src/main/resources/JavaSpring/model.mustache (hunk 2 above)
Comment (welshm) [inline modules/openapi-generator/src/main/resources/JavaSpring/model.mustache:49]: Is this still necessary after https://github.com/OpenAPITools/openapi-generator/pull/11784 ?

**Comment by cachescrubber** [comment]
@wing328. Thanks for the review and merging!

**Pull request merged**